### PR TITLE
[New arch] Create folder

### PR DESCRIPTION
**Pull request by abelgardep**
Implements https://github.com/owncloud/android/issues/2861

### CODE REVIEW

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        var result = createFolder(client)
+        if (!result.isSuccess && createFullPath && result.code == ResultCode.CONFLICT) {
+            result = createParentFolder(FileUtils.getParentPath(remotePath), client)
+
+            if (result.isSuccess) {
+                result = createFolder(client)
+            }
+        }
+        return result
+    }
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:63]: Reusing result variable here makes the code less readable, could you write it in a different way?

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    private fun createFolder(client: OwnCloudClient): RemoteOperationResult<Unit> {
+        var result: RemoteOperationResult<Unit>
+        try {
+            val webDavUri = if (isChunksFolder) client.uploadsWebDavUri else client.userFilesWebDavUri
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:68]: I would use braces `if (isChunksFolder) { client.uploadsWebDavUri } else { client.userFilesWebDavUri }` but up to you 🍻

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            val mkcol = MkColMethod(URL(webDavUri.toString() + WebdavUtils.encodePath(remotePath))).apply {
+                setReadTimeout(READ_TIMEOUT, TimeUnit.SECONDS)
+                setConnectionTimeout(CONNECTION_TIMEOUT, TimeUnit.SECONDS)
+            }
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:72]:
```suggestion
            val mkcol = MkColMethod(
                URL(webDavUri.toString() + WebdavUtils.encodePath(remotePath))
            ).apply {
                setReadTimeout(READ_TIMEOUT, TimeUnit.SECONDS)
                setConnectionTimeout(CONNECTION_TIMEOUT, TimeUnit.SECONDS)
            }
```

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/services/FileService.kt [hunk 4]
```diff
@@ -29,5 +29,6 @@ import com.owncloud.android.lib.resources.files.RemoteFile
 
 interface FileService: Service {
     fun checkPathExistence(path: String, isUserLogged: Boolean): RemoteOperationResult<Boolean>
+    fun createFolder(remotePath: String, createFullPath: Boolean, isChunkFolder: Boolean = false): RemoteOperationResult<Unit>
```
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/services/FileService.kt:32]:
```suggestion
    fun createFolder(
        remotePath: String, 
        createFullPath: Boolean, 
        isChunkFolder: Boolean = false
    ): RemoteOperationResult<Unit>
```

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt (hunk 2 above)
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:68]: I agree 🍻

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt (hunk 3 above)
Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:72]: 💯

File: owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt (hunk 1 above)
Comment (davigonz) [inline owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:63]:
```suggestion
override fun run(client: OwnCloudClient): RemoteOperationResult<Unit> =
    createFolder(client).takeUnless { 
        !it.isSuccess && createFullPath && it.code == ResultCode.CONFLICT 
    } ?: createParentFolder(FileUtils.getParentPath(remotePath), client).takeUnless { 
        it.isSuccess 
    } ?: createFolder(client)
```

Comment (abelgardep) [reply owncloudComLibrary/src/main/java/com/owncloud/android/lib/resources/files/CreateRemoteFolderOperation.kt:63]: I'll keep the current behavior at least for the moment. Thanks anyway 👍

**Pull request merged**